Better error reporting in normalize*

holyjak opened the pull request:
First step on the way to fixing #491

holyjak commented:
I have improved the error messagefrom #491  to be the following:

> Query join indicates the data should contain a data map but the actual data is "Important" Joined component's ident: [:tag/id nil]

holyjak commented:
If anybody knows of any other cases where normalize fails with an unhelpful error message, pls let me know!

awkay commented on src/main/com/fulcrologic/fulcro/algorithms/normalize.cljc:
```diff
@@ -18,74 +20,193 @@
+    (normalize* query data tables union-seen transform)
+    (catch #?(:clj Exception :cljs :default) e
+      (normalize-with-friendly-errors query data tables union-seen transform)
+      (log/warn "Unexpectedly, normalize* failed but normalize-with-friendly-errors did not. Please report it (with the inputs) to Fulcro")
```
I would think the ideal thing to do would be report error, then issue a final error that says "normalize failed, see above errors" and maybe includes the query and data that was provided...but don't throw again. A stack trace is almost certainly non-helpful since it will come from the low-level transaction system, not your code.

holyjak replied:
Ok! Pushed a new version.

I still throw an exception in `upsert-ident` just to fail fast (and get the info faster to the developer) but changed this place to only log.

awkay commented:
Where were we on this??? I thought it was merged already :)

holyjak commented:
No, I had not time to work on the suggested changes due to re:Clojure. I will get back to it ASAP.

holyjak commented:
Reproducing from Slack:

> [Tony, Nov 15th]: My apologies on the normalize one. I’m torn. If you were only after improving one particular case of potential error, then perhaps it does make more sense just to augment the original function. I’m concerned about breakage, so the zipmap should remain as-is, but I think it is possible to refactor in some error checking without breaking things.  I had assumed you wanted to do a bunch of things to check for various errors (not sure what those would be off the top of my head, it was just my assumption).  What’s your take on the fragility of just adding this one error message to the original function?

holyjak commented:
### Testing

Test code, to be inserted into `com.fulcrologic.fulcro.algorithms.normalize`:

```clojure
(comment
  (do
    ;; FIXME: Why does this often take > 5s to run? Seems caused by `log/error in `upsert-ident, even if it only logs "test" ?!
    (comp/defsc Tag [_ _]
      {:query [:tag/id :tag/desc]
       :ident :tag/id
       :initial-state {}}
      nil)

    (comp/defsc Menu [_ _]
      {:query [{:tags (comp/get-query Tag)}] ; <-- WRONG, tags are strings not maps
       :ident (fn [] [:component/id :Menu])
       :initial-state {}}
      nil)

    (comp/defsc Root [_ _]
      {:query [{:task-filters (comp/get-query Menu)}]
       :initial-state (fn [_] {:task-filters {:tags ["Important" "Urgent"]}})}
      nil)

    (tree->db Root {:task-filters {:tags ["Important" "Urgent"]}}))
  )
```

The output will be:

> ERROR - Normalize failed and no data will be inserted into the client DB. Error: Query join indicates the data should contain a data map but the actual data is "Urgent" Joined component's ident: [:tag/id nil] Target component: com.fulcrologic.fulcro.algorithms.normalize/Root Data:  {:task-filters {:tags ["Important" "Urgent"]}}
>
> `=> {}`

holyjak commented:
### What has been done

1. Throw a user-friendly exception instead of the one thrown previously
2. Moved `update-in ...` into the friendly named `upsert-ident`
3. Renamed locals to more meaningful names
4. Added info about the origin of the code so that we know it if/when we consider refactoring in the future

holyjak commented:
From the [history of my discussions with Tony](https://clojurians.slack.com/archives/D010CPAPAH1/p1636145990010300):

> tony.kay  21 h 59 Nov 5th
>
> So, it would be nice if normalize could continue. For example, if you got back `m` without the `update-in` applied you'd just skip bad data in that case so, I wonder if making some kind of guarded wrappers for assoc/assoc-in/update/update-in would be better
>
> Jakub Holý: For the developer it is "better" if it blows up, with a good error, b/c then s/he will notice the problem 
>
> Tony: I don't agree with that [an error blowing up the ui]. An error in the logs that explains why my data isn't in state is better to me. I don't need a stack trace into an algorithm I didn't even directly call. I dislike the "burn the UI down" behavior of react in general...so I'm not a fan of making it happen ever
